fastMatcher.go

```diff
@@ -625,6 +626,33 @@ func (m *FastMatcher) matchExec(token tokenType, tokenData []byte, tokenDataLen
 	return nil
 }
 
+// If any operation is done strictly on the key of an obj or array, this method takes care of it
+func (m *FastMatcher) matchKeyOfObj(token tokenType, tokenData []byte, node *ExecNode) (error, bool) {
+	lastArrOrMapKeyByteLen := len(m.lastArrOrMapKeyBytes)
+	if lastArrOrMapKeyByteLen == 0 {
+		return nil, false
+	}
+
+	var savedKey []byte = m.lastArrOrMapKeyBytes
+	m.lastArrOrMapKeyBytes = nil
+	if len(node.Ops) == 0 || (token != tknObjectStart && token != tknArrayStart) {
+		return nil, false
+	}
+
+	// Since this matchExec should only execute on the specific key
+	// temporarily don't do any afters, since they are out of scope
```

Comment:
https://github.com/couchbaselabs/gojsonsm#afters is an explanation.
So if an after node exists on the specific execNode, it means more than likely that the KV's in the embedded obj or arrays belong there, and whatever op is only done on the key belongs in that same level of execNode.
@brett19  can fact-check me.

Reply:
After blocks are executed AFTER an entire tree has been processed.  Imagine something like `{"z": {"a":1, "b":2, "c":3}}`, if you have an after on `z`, it will be processed once the whole object has already been processed.  This is used in cases where you have a condition such as `z.a=z.b`.  In these cases we must have both `z.a` and `z.b` in scope at the same time, JSONSM stores a pointer to the value of `z.a` in $1 when it passes it, then stores a pointer to `z.b` in $2 when it processes that.  Finally, in the after block for `z` it can execute something like `$1=$2` in order to check whether the original `z.a=z.b` passed.

Reply:
i take it that we should not set node.after to nil, then?

Reply:
Brett mentioned that nil after token is acceptable. The question here is, if node.after happens not to be nil, and we treat it as nil in here, would something break.

Reply:
The concern here is that if the after token is temporarily set to nil, would the content of the after tokens missed and not be evaluated, causing an incorrect output.
The mechanism that determines when it is safe to return, aka The "shortcut" evaluation, is captured by the means of the binTree. If the binTree at root node 0 is determined that the whole binTree has been evaluated, then the whole evaluation is considered finished. 
An expression that is of this concern can be thought of as `EXISTS(z) AND z.a != z.b`
When the transformer transforms this, the binTree will be represented as
```[0] and
  [1] leaf
  [2] not
    [3] leaf
```
0 is the root node, which is used to evaluate if all the rest are evaluated
1 is the node that represents `EXISTS(z)`
2 is the NOT op
3 is the op of `z.a = z.b`

Separate from the binTree is the ExecNode structure, in which it looks like:

```
  `z`:
    :ops
      [1] @ exists @
    :elems
      `a`:
        :store $5
      `b`:
        :store $6
    :after-ops:
      [3] $5 eq $6
```
With the code addition I added, JsonTokenizer will read the json document and sees the key of "z", and perform the `matchOp` on `z`. Note this will be a one-time evaluation and not recursive, because the key *must* be a literal and not an object start. If the after-ops are present, it will evaluate the after-ops at this point and cause an invalid evaluation (because $5 and $6 hasn't been read yet)

When the key of `z` is evaluated, the matcher will ask the binTree structure if the whole tree has been resolved (for short circuit purposes). It hasn't, because `z.a` and `z.b` hasn't been marked yet. So this will allow the matchOp on the key to return and afterNodes will be restored.

Once the key itself has been evaluated, the after ops are restored, and those will then be evaluated after `z.a` and `z.b` has been gathered. The same code path remains untouched.